cinn/frontend/pass/transpose_collapsing.cc

```diff
@@ -0,0 +1,349 @@
+#include <unordered_set>
+
+#include "cinn/common/target.h"
+#include "cinn/frontend/cinn_builder.h"
+#include "cinn/frontend/program_pass.h"
+#include "cinn/frontend/syntax.h"
+
+namespace cinn::frontend::pass {
+
+using cinn::utils::DimType;
+using cinn::utils::ShapeType;
+
+using OutputToOpMap = std::unordered_map<std::string, Instruction*>;
+using InputToOpMap  = std::unordered_map<std::string, std::unordered_set<Instruction*>>;
+
+// Pass `TransposeCollapsing` folds multi transpose into one.
+class TransposeCollapsingPass : public ProgramPass {
+ public:
+  using ProgramPass::ProgramPass;
+
+ protected:
+  void ApplyImpl(Program* program,
+                 const std::unordered_set<std::string>& fetch_ids,
+                 const common::Target& target) const override {
+    VLOG(4) << "Run TransposeCollapsingPass";
+    // `out2instr` is used to represent the mapping of Output to Instruction.
+    OutputToOpMap out2instr;
+    // `in2instr` is used to represent the mapping of Input to Instruction.
+    InputToOpMap in2instr;
+    // all transpose op in program
+    std::unordered_set<Instruction*> all_transpose;
+
+    for (size_t i = 0; i < program->size(); ++i) {
+      auto& instr = (*program)[i];
+      for (const auto& out : instr->outputs) {
+        out2instr[out->id] = &instr;
+      }
+      for (const auto& in : instr->inputs) {
+        in2instr[in->id].insert(&instr);
+      }
+      if ("transpose" == instr->op_type) {
+        all_transpose.insert(&instr);
+      }
+    }
+
+    // the useless transpose op need to remove from program
+    std::unordered_set<Instruction*> remove_instrs;
+    FoldingTransposeVertical(all_transpose, fetch_ids, in2instr, out2instr, &remove_instrs);
+
+    for (auto instr : remove_instrs) {
+      if (all_transpose.count(instr)) {
+        all_transpose.erase(instr);
+      }
+    }
+    FoldingTransposeHorizontal(all_transpose, fetch_ids, in2instr, out2instr, &remove_instrs);
+
+    CinnBuilder builder("transpose_collapsing_builder");
+    for (auto& var : program->GetInputs()) {
+      builder.CreateInput(var);
+    }
+    for (int i = 0; i < program->size(); i++) {
+      if (remove_instrs.end() == remove_instrs.find(&(*program)[i])) {
+        builder.AppendInstruction((*program)[i]);
+      }
+    }
+    *program = builder.Build();
+  }
+
+ private:
+  static void FoldingTransposeVertical(const std::unordered_set<Instruction*>& all_transpose,
+                                       const std::unordered_set<std::string>& fetch_ids,
+                                       const InputToOpMap& in2instr,
+                                       const OutputToOpMap& out2instr,
+                                       std::unordered_set<Instruction*>* remove_instrs) {
+    // the transpose op should not remove
+    std::unordered_set<Instruction*> visited_instrs;
+    for (auto transpose : all_transpose) {
+      if (remove_instrs->count(transpose) || visited_instrs.count(transpose)) {
```

Comment:
这里好像只需要使用visited_instrs即可？
```suggestion
      if (visited_instrs.count(transpose)) {
```

Reply:
之前有个方案是创建新的instruction而不是利用现有的第二个transpose，所以需要判断下remove_instrs。后来改了，的确只用判断visited_instrs就行了。Done